networks/rpc/handler.go

```diff
@@ -0,0 +1,451 @@
+		h.handleCall(ctx, msg)
+		logger.Debug("Served "+msg.Method, "duration", time.Since(start))
+		return nil
+	case msg.isCall():
+		resp := h.handleCall(ctx, msg)
+		if resp.Error != nil {
+			logger.Warn("Served "+msg.Method, "reqid", idForLog{msg.ID}, "duration", time.Since(start), "err", resp.Error.Message)
+		} else {
+			logger.Debug("Served "+msg.Method, "reqid", idForLog{msg.ID}, "duration", time.Since(start))
+		}
+		return resp
+	case msg.hasValidID():
+		return msg.errorResponse(&invalidRequestError{"invalid request"})
+	default:
+		return errorMessage(&invalidRequestError{"invalid request"})
+	}
+}
+
+// handleCall processes method calls.
+func (h *handler) handleCall(cp *callProc, msg *jsonrpcMessage) *jsonrpcMessage {
+	if msg.isSubscribe() {
+		return h.handleSubscribe(cp, msg)
+	}
+	var callb *callback
+	if msg.isUnsubscribe() {
+		callb = h.unsubscribeCb
+		wsUnsubscriptionReqCounter.Inc(1)
+	} else {
+		callb = h.reg.callback(msg.Method)
+	}
+	if callb == nil {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&methodNotFoundError{method: msg.Method})
+	}
+	args, err := parsePositionalArguments(msg.Params, callb.argTypes)
+	if err != nil {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&invalidParamsError{err.Error()})
+	}
+	return h.runMethod(cp.ctx, msg, callb, args)
+}
+
+// handleSubscribe processes *_subscribe method calls.
+func (h *handler) handleSubscribe(cp *callProc, msg *jsonrpcMessage) *jsonrpcMessage {
+	if !h.allowSubscribe {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(ErrNotificationsUnsupported)
+	}
+
+	if int32(len(h.serverSubs)) >= MaxSubscriptionPerWSConn {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&callbackError{
+			fmt.Sprintf("Maximum %d subscriptions are allowed for a websocket connection. "+
+				"The limit can be updated with 'admin_setMaxSubscriptionPerWSConn' API", MaxSubscriptionPerWSConn),
+		})
+	}
+
+	// Subscription method name is first argument.
+	name, err := parseSubscriptionName(msg.Params)
+	if err != nil {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&invalidParamsError{err.Error()})
+	}
+	namespace := msg.namespace()
+	callb := h.reg.subscription(namespace, name)
+	if callb == nil {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&subscriptionNotFoundError{namespace, name})
+	}
+
+	// Parse subscription name arg too, but remove it before calling the callback.
+	argTypes := append([]reflect.Type{stringType}, callb.argTypes...)
+	args, err := parsePositionalArguments(msg.Params, argTypes)
+	if err != nil {
+		rpcErrorResponsesCounter.Inc(1)
+		return msg.errorResponse(&invalidParamsError{err.Error()})
+	}
+	args = args[1:]
+
+	// Install notifier in context so the subscription handler can find it.
+	n := &Notifier{h: h, namespace: namespace}
+	cp.notifiers = append(cp.notifiers, n)
+	ctx := context.WithValue(cp.ctx, notifierKey{}, n)
+
+	rpcSuccessResponsesCounter.Inc(1)
```

Comment:
we may remove rpcSuccessResponsesCounter.Inc(1) here.

Reply:
thx!
resolved by 0d4ea1d